### PR TITLE
fine-grained errors reporting for PaymentAddress

### LINKED ISSUE #647: Fine grained error recovery for fields
[Similarly to Apple Pay](https://developer.apple.com/documentation/applepayjs/applepayerrorcontactfield), we need need to define a means to signal which fields in the payment sheet suffer from a validation error.

### PR DESCRIPTION
**Pull request by marcoscaceres**
closes #647
 
The following tasks have been completed:

 * [x] Confirmed there are no ReSpec errors/warnings.
 * [x] [Added Web platform tests](https://github.com/web-platform-tests/wpt/pull/12396)
 * [ ] added MDN Docs (link)

Implementation commitment:

 * [x] Safari - already implemented.
 * [ ] Chrome (link to issue)
 * [x] [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1462907)
 * [ ] Edge (public signal)

Impact on Payment Handler spec?

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/payment-request/pull/712.html" title="Last updated on Nov 2, 2018, 9:58 PM GMT (a9aee3f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/payment-request/712/00a0976...a9aee3f.html" title="Last updated on Nov 2, 2018, 9:58 PM GMT (a9aee3f)">Diff</a>

### CODE REVIEW

**Comment by marcoscaceres** [comment]
First draft ready for review. This part only deals with errors related to shipping addresses. 

Will start working on #705, which will then give us the ability to correct both address-releated errors and Payer Errors (e.g., bad email) in the `PaymentResponse`.

File: index.html
Comment (marcoscaceres) [inline index.html]: Kept this (languageCodeError) for completeness... but wonder if we should just ditch languageCodeError. 🤔

File: index.html
Comment (marcoscaceres) [inline index.html:2756]: Decided to postfix all these with "Error" in the name, as I felt it made it a little bit more clear when in userland code.

File: index.html
Comment (marcoscaceres) [reply index.html]: Clarified what "this" was above.

**Comment by marcoscaceres** [comment]
We have a partial implementation of this in Firefox now (currently based on the original proposal - but will get updated to match the new one soon).

File: index.html
Comment (msujaws) [reply index.html:2756]: The extra "Error" suffix feels redundant me, as it should be implied since these are members of the shippingAddressErrors property.

Comment (marcoscaceres) [reply index.html:2756]: depends... the problem I ran into was in destructuring and using object short hands. Also, consider the `shippingAddressErrors` can be the result of another function. Thus:

```JS
function validateAddress(address){
   // whoops, now "postalCode" has been squatted. 
   const  { postalCode } = address;
   const postalCodeError =  !isValidPostCode(postCode) ? "Invalid zip code" : undefined;
   return { postalCodeError }
}

function collectErrors(response) {
   // same problem here!
   const { shippingAddress } = response;
   const  shippingAddressErrors = validateAddress(shippingAddress);
   return { shippingAddressErrors }
}
```

And so on...

Comment (msujaws) [reply index.html:2756]: Fair enough :)

**Comment by marcoscaceres** [comment]
@domenic, when you have a few mins, would appreciate a final review. We've landed this in Firefox.

File: index.html
Comment (domenic) [reply index.html:2756]: I tend to agree this is redundant. I think you have chosen a code style with a combination of object literal shorthand, destructuring, and redundant variable declarations that makes it extra confusing. But I don't think that's a good reason to verbose-ify the API, personally. I would instead suggest writing your code in a more conventional way, such as

```js
function validateAddress(address){
   return {
    postalCode: !isValidPostalCode(address.postalCode) ? "Invalid zip code" : undefined
  };
}

function collectErrors(response) {
   return {
    shippingAddress: validateAddress(response.shippingAddress)
  };
}
```

(basically, don't declare variables you're only going to use once.

File: index.html
Comment (domenic) [inline index.html]: I'd say "firstly, its presence denotes..."

File: index.html
Comment (domenic) [inline index.html:522]: Empty shippingOptions should not be necessary, since not-present gets converted to empty in the processing model, I am pretty sure.

**Review by domenic — APPROVED** [review]
Mostly LGTM, although I agree with others who are concerned about the redundancy. Will leave it to you to make the call though.

File: index.html
Comment (marcoscaceres) [reply index.html:522]: In the "[update a PaymentRequest's details algorithm](https://www.w3.org/TR/payment-request/#dfn-update-a-paymentrequest-s-details-algorithm)", we do an incremental update to `request`, so:

> If the `shippingOptions` member of details is present, and `request.[[options]].requestShipping` is true, then:
>  1. Set request.[[details]].shippingOptions to shippingOptions.

Otherwise, the original shipping options are retained - so the developer needs to be explicitly thrash them.

**Comment by marcoscaceres** [comment]
@ianbjacobs, could you please republish the document on TR?

**Comment by ianbjacobs** [comment]
Hi @marcoscaceres, I have send a publication request for 7 June. Thanks for all the editing!

**Comment by marcoscaceres** [comment]
Added tests https://github.com/web-platform-tests/wpt/pull/12396

**Pull request merged**
